### PR TITLE
MGDOBR-1037: Shard should verify hostname and SSL configuration of the Route exposed for a Bridge as part of the reconcile loop

### PR DESCRIPTION
**Pull request by manstis**
[https://issues.redhat.com/browse/MGDOBR-1037](https://issues.redhat.com/browse/MGDOBR-1037)

It's not clear what `RouteSpec` properties we should be considering in the comparison.

However, since `RouteSpec` overrides `equals(..)` delegate complete comparison to it.

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/OpenshiftNetworkingService.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ public NetworkResource fetchOrCreateBrokerNetworkIngress(BridgeIngress bridgeIng
                 .withName(bridgeIngress.getMetadata().getName())
                 .get();
 
-        if (existing == null || !expected.getSpec().getTo().getName().equals(existing.getSpec().getTo().getName())) {
+        if (existing == null || !existing.getSpec().equals(expected.getSpec())) {
```
Comment (sutaakar) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/OpenshiftNetworkingService.java:55]: OpenShift adjusts Route spec definition with default parameters if not defied (i.e. host, to.weight, wildcardPolicy).
Such change will IMHO cause infinite reconciliation.
I think we need to check specific parameters.

Comment (manstis) [reply shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/OpenshiftNetworkingService.java:55]: @sutaakar Is there a definitive set of fields we should check?

We used to compare `To` as a whole (and not properties excluding `to.weight` so we have a problem already.

I really could not find anything to suggest what specific properties we should check.

Even `To` (as defined in our template) is, as you write, not entirely comparable.

Comment (sutaakar) [reply shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/OpenshiftNetworkingService.java:55]: In previous version I see that we compare just `to.name`.

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/OpenshiftRouteSpecMatchesHelper.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    private OpenshiftRouteSpecMatchesHelper() {
+        //Static utility class
+    }
+
+    public static boolean matches(RouteSpec existing, RouteSpec expected) {
+        Function<RouteSpec, RouteSpec> spec = (RouteSpec source) -> source;
+        if (!Objects.equals(spec.apply(existing), spec.apply(expected))) {
+            return false;
+        }
```
Comment (sutaakar) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/OpenshiftRouteSpecMatchesHelper.java:19]: This check returns false even when existing and expected don't match in a parameter which we don't check (i.e. `to.weight`).

Comment (manstis) [reply shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/OpenshiftRouteSpecMatchesHelper.java]: @sutaakar Fixed.. I made a right mess of this change. Hopefully it's OK now.

**Review by sutaakar — APPROVED** [review]
Looks good to me

**Review by r00ta — APPROVED** [review]

**Review by Srihari1192 — APPROVED** [review]

**Pull request merged**